### ISSUE TITLE
Distribute B3 as Frozen application

### ISSUE THREAD

**Issue by danielepantaleone** [BigBrotherBot/big-brother-bot#254]
I'm opening this issue so we can keep talking about distributing B3 as a Frozen application as being discussed in #253 . I have been thinking a lot about this in the last couple of days and I believe that for end users will be way easier to install and run B3 if it would be composed of a single binary file + configuration files:

**PROS**
* Easy install: B3 will be python indipendent since python sources will be compiled into a single binary file (like it's being done for the Win32 version of B3)
* We could drop python 2.6 support (which is not required but will cleanup a lot B3 source code by removing all the `try ... except` needed to load the correct library, and we can also remove some of the libraries we are shipping in the `lib` directory)
* Easy update: manually drop the new binary file on the old one (overwriting it). the updater plugin will be simplified a lot, since it would need to take care of 1 file only
* Reduce support requests: we won't have to deal with users not being able to install the correct python version, or people that while upgrading B3 to a newer version forgot to update a python module.

**CONS**
* For every release (including daily builds) we would need to compile B3 as Frozen application at least for Linux, OSX and Windows. We would need to setup an automated procedure that does the compile automatically (and so @thomasleveil won't have to manually take care of pushing daily builds every now and then).
* Every compile needs to be done on the specific operating system (and this may generate some problems, but nothing which can't be solved)

Obviously there will still be the possibility of running B3 from sources.
Thoughts?

**Comment by thomasleveil** [comment]
The only thing we will _loose_ is support for python 2.6, but AFAIK we were only keeping it to please users of CentOS. I now see that CentOS 7 ships with python 2.7, so it looks like we can drop support for python 2.6 at last ;)

**Comment by danielepantaleone** [comment]
We lose support for python 2.6 on old version of CentOS only for people that run B3 from sources. If we provided a compiled binary it should run on old versions of CentOS

**Comment by thomasleveil** [comment]
A friend of mine advised [PEX](https://pex.readthedocs.org/en/latest/). I'm not sure this would be better than frozing B3 but still worth the look

**Comment by danielepantaleone** [comment]
I had a look and I don't think it's suitable for B3. Also as far as I understood it's supported only on Linux systems (or maybe I skipped the Windows part).

I still think that the easiest way is distributing B3 as single binary application together with all the files needed: something like **unzip and run**.

**Comment by danielepantaleone** [comment]
I would like to have some more opinions on the matter from @markweirath @ozguruysal @ozon @82ndab-Bravo17 so we can either move forward in this direction or close the issue.

**Comment by markweirath** [comment]
First we have the source code for ppl with a python interpreter installed and some advanced knowledge of installing B3 from source and its dependencies. These installation have most options like shared source code for multiple bots and such.

We then made that easier for linux users using easy install and pypi (.egg files)

After that we wanted to ease up the installation process for windows users by creating binary packages for windows.

So the next step is to ease up the installation process for linux users by freezing B3 into a single package?

To summarize this we would distribute these packages:
- source code (zipped)
- linux .egg
- windows binary installer
- windows binary standalone
+ linux frozen package

I see the frozen package as a competitor to the .egg package and if it makes upgrading the linux-frozen and windows-binary easier I see that as an advantage. I have no idea how many users are using the pypip packages to upgrade their bots... so no idea how many advanced users will be affected.

**Comment by danielepantaleone** [comment]
My idea is:

* source code zipped
* egg distribution
* Windows binary installer (dunno how much is necessary)
* Windows binary standalone
* Linux binary standalone
* OSX app

The reason behind all this is that it may seem easy to setup and start B3 if you have some python basic knowledge but there are some things which needs to be done before you have B3 running:

* install python if your OS doesn't ship it
* install the correct version of python
* install the required libraries
* figure out how to run B3 in background (mostly used on linux systems)
* ....

I solved some of those problems by shipping together with B3 a Linux bash script (https://github.com/danielepantaleone/big-brother-bot/tree/release-1.10/scripts) which ease the B3 setup (by checking python version, adding capabilities of starting B3 in background, starting multiple B3 using the same source, etc...) but I still believe we can ease it more.

I'm trying to look at the picture from a common user perspective: when I download a software I expect to use it in the next few minutes, something like:

- download
- setup
- launch

Distributing a Frozen application will solve all those matters since:

* python will be shipped within the binary (together with all the necessary libraries)
* you would have to deal with just one binary file, so the update process will be really easy (just replace it)
* the setup procedure starts automatically when no b3.ini is found so I see this like: drop somewhere and launch, then you are good to go.

In the same way, I suggested (and then implemented) a plugin module structure change (https://github.com/BigBrotherBot/big-brother-bot/issues/247). Installing a plugin was way too complicated for the end user (infact we had a lot of Forum posts with embedded b3.log showing multiple external plugins not being loaded, and thus crashing B3, because of a missing configuration file). I wanted something easy like, again, drop the plugin folder inside the extplugins folder, add a line in b3.ini (also without specifying the config file since B3 now search it automatically) and then reboot B3.

Basically I would like to achieve the same for B3 core (despite the fact that advanced users, including me, will keep using the source version)

**Comment by thomasleveil** [comment]
Here are the download stats on the last 30 day for the daily builds I host:
![image](https://cloud.githubusercontent.com/assets/42300/6096854/cbd48192-afa6-11e4-884e-d8e39763387b.png)

very few users are using egg. Those 8 download could have even been made by error. I will now on stop to build the egg file for daily builds and we'll see if anyone complains.

**Comment by ozguruysal** [comment]
I think any step to make things easier for end users is a good step. Source files (stable or development) are always available in GitHub for advanced users. I would make things more simple and provide:

* source code zipped
* Windows binary standalone
* Linux binary standalone

I don't think egg distribution or windows binary installer would be needed in this case. Also I'm not sure if we have any OSX users.

Thanks @thomasleveil for providing those stats. They are showing us that we should consider Windows users more these days with all the recent games lacking linux server support. And for windows users it's always click and run.

**Comment by thomasleveil** [comment]
Keep in mind those stats are for daily builds and as such target users who are willing to stay on the edge. I guess such users tend be more resourceful and independent and that might explain why there are fewer downloads of the win32 installer.
However upon releasing 1.10 as stable or for 1st time users I would guess the win32 installer package will be more popular ; simply because it provides a "walk through" and also such installers are using well known idioms which is reassuring for 1st timers.

**Comment by danielepantaleone** [comment]
I created a new branch: https://github.com/danielepantaleone/big-brother-bot/tree/cx_freeze
I added there a `build.py` script which packages the win32 zip distribution of B3 using `cx_Freeze` (which has support also for OSX and Linux distributions).
I'm not very experienced in using setuptools so I would like some opinions/review from @thomasleveil (for example I don't know how to integrate this with the .egg distribution package generation).

**Comment by thomasleveil** [comment]
I'm not uptodate with the new ways to properly distribute python code. When you read [this SO response](http://stackoverflow.com/questions/6344076/differences-between-distribute-distutils-setuptools-and-distutils2/14753678#14753678) you wonder what the hell you're supposed to do ^^

In B3 we are using [stuff that dates from 2008](https://github.com/BigBrotherBot/big-brother-bot/blob/master/ez_setup.py) and we might want to dust off that as well.

**Comment by danielepantaleone** [comment]
Which is why I said "I'm not very experienced", since I wasn't here in 2008 and I don't know what is needed, for which platform, and why things have been added. If I have to do it "my way" I will simply remove everything and redo `setup.py` to build:

* source package
* win32 zip
* linux zip
* OSX zip (well, since cx_Freeze can do it, why not ^^)

One thing that I noticed is that differently form py2exe cx_Freeze creates a lot of files to be shipped with the executables, while py2exe does not (but I don't think it's a problem).

**Comment by danielepantaleone** [comment]
Some changes done in my [cx_Freeze](https://github.com/danielepantaleone/big-brother-bot/tree/cx_freeze) branch:

* removed setuptools in favor of distutils (cx_Freeze can't work with setuptools)
* refactored `setup.py` to build `source` and `binary` distribution (according to the OS running: so far I tested only Windows, but I plan to setup a couple of virtual machines tomorrow to test also OSX and Linux)
* removed py2exe support
* removed (for the moment) .egg generation (with distutils it's kinda tricky so I would have to look deeper in the documentation)

**Comment by 82ndab-Bravo17** [comment]
I would suggest keeping the win32 installer, I would think that it is  the one most likely to be used by first time users and it does do a nice job of being a fairly simple installation process. After the first install people could update using the zip package, but as we saw recently even that is not a foolproof process!

**Comment by thomasleveil** [comment]
Same feeling here, the windows installer still have a purpose.

@danielepantaleone can you give the different commands to run to make the different builds?

**Comment by danielepantaleone** [comment]
Alright, pushed some other changes which add compile support for linux. 
So basically `cx_Freeze` makes use of `distutils.core.setup` which is normally used to package python sources. I made use only of `cx_Freeze.setup` to package both source distribution and a compiled distribution (compiled according to the operating system running at build time). The only command I implemented so far is `python setup.py build` which produces 2 files on my Windows machine (both located in the build directory):

* `b3-1.10.0dev-20150212.zip` (source distribution)
* `b3-1.10.0dev-20150212-win32.zip` (because I ran the build on linux)

I also tested the linux build (both using 32bit python and 64 bit python and they works perfectly. The InnoSetup script would need to be adapted if we want to keep the Windows installer and also the .egg distribution is also missing.

**Comment by thomasleveil** [comment]
There is also another thing to figure out: would that source distribution be suitable for uploading to [Pypi](https://pypi.python.org/pypi/b3/) and would a tools such as fig be able to install B3?

**Comment by danielepantaleone** [comment]
Those things I still have to understand (as I told u I'm not experienced in packaging but I'm learning so far :smile:) The source distribution is the very same being done previously, except that now I used `distutils` instead of `setuptools` (I was using setuptools in combination with `cx_Freeze` but setuptools overrides some variables defined in distutils (used by `cx_Freeze`) and so the build was failing)

**Comment by thomasleveil** [comment]
I'm reading the [Python Packaging User Guide](https://python-packaging-user-guide.readthedocs.org/en/latest/) and it seems the new way of publishing to pypi is by using [wheel](http://pythonwheels.com/) distribution (which is intended to replace eggs).

So I guess our next step is: building a wheel distribution

**Comment by danielepantaleone** [comment]
About the error in Jankins I think it's a pretty recent bug: https://bitbucket.org/anthony_tuininga/cx_freeze/issue/132/cx_freezeutil-dll-load-failed-the
However it compiles fine on my system but I'm using python 32 bit on Windows. What version of python Jenkins use?

**Comment by thomasleveil** [comment]
> jenkins

still failing: http://ci.cucurb.net/view/Dist/job/B3-1.10-danielepantaleone-cx_freeze/9/console

**Comment by danielepantaleone** [comment]
Also it seems there is no way of building wheele and egg using distutils: all the documentation I found is related to setuptools. I started using cx_freeze because with the same script is possible to generate executables for Windows, OSX and Linux, but probably I should look into integrating py2app and Freeze (for linux) in our setuptools script

**Comment by danielepantaleone** [comment]
I rebased everything using setuptools + cx_freeze combination in a new branch [setuptools](https://github.com/danielepantaleone/big-brother-bot/tree/setuptools)
I adapted the `setup.py` script to build source distribution, python .egg and platform binary.
To make a full build:

`python setup.py build_exe egg_info --tag-build .daily317 --tag-date sdist bdist_egg`

The build_exe command (also just 'build' its fine since I aliased it in sys.argv) needs to be placed at the beginning otherwise when the `bdist_egg` command gets executed, which will replace the version number using the tag_build and tag_date, will conflict with `cx_Freeze`. This is my current output:

```
BUILD: C:\Users\Daniele\workspace\big-brother-bot\build\b3-1.10.0.daily317-20150213-win32.zip
BUILD: C:\Users\Daniele\workspace\big-brother-bot\build\b3-1.10.0.daily317-20150213.zip
BUILD: C:\Users\Daniele\workspace\big-brother-bot\build\b3-1.10.0.daily317_20150213-py2.7.egg
```

@thomasleveil could you try to build it on Jenkins again? Also if it fails with the same error could you try the previous version of cx_Freeze or using the 5.0 development snapshot?

**Comment by danielepantaleone** [comment]
Another update: I changed the `setup.py` script and adapted the `setup.cfg` configuration file to build the following:

* source distribution (.zip and .tar.gz formats)
* python .egg distribution
* compiled binary (standalone)
* compiled binary (windows installer)

The command I tested so far (and which works perfectly) are:

```
python setup.py build_exe egg_info --tag-build .daily999 --tag-date sdist bdist_egg
python setup.py release
python setup.py alpha
python setup.py beta
python setup.py dev
python setup.py prepare
python setup.py test
```

Again, Windows and Linux compile is supported: as for OSX I would have to power on my MacBook and configure it with all the missing libraries to give it a try (and I plan to do it next week). 
@thomasleveil I saw u ran a build with Jenkins and it failed. Any clue?

**Comment by thomasleveil** [comment]
no clue yet :/ I couldn't give sufficient time to investigate either

**Comment by danielepantaleone** [comment]
No worries :smile: was just curious.
Anyway, to me, this seems to be going quite good. The only difference with the `py2exe` version of `setup.py` is that `cx_Freeze` generate a lot more files, while `py2exe` package everything into `b3.lib` but I don't see this as a problem (at least looking at the fact that we will benefit from having compiled binaries for all the major platforms).

**Comment by thomasleveil** [comment]
I don't see that as an issue either

**Comment by thomasleveil** [comment]
Back on Jenkins: different error now [`ImportError: No module named 'irc'`](http://ci.cucurb.net/job/B3-1.10-danielepantaleone-setuptools/12/console)

For the record, what I did so far:
1. install cx_freeze from https://bitbucket.org/anthony_tuininga/cx_freeze/get/tip.tar.gz
1. install _Microsoft Visual C++ 2008 SP1 Redistributable Package (x86)_ (but I'm not sure that was needed)
1. install _Microsoft Visual C++ Compiler for Python 2.7_ http://aka.ms/vcpython27 and followed carefully the install instructions to install if for all users

**Comment by danielepantaleone** [comment]
The irc is a python module I added to the requirements of cx_Freeze because it's needed by an external plugin I made (which connects B3 to an IRC server). Since the irc library is being updated quite frequently, I didn't placed it withing the plugin. To install it just do `pip install irc`. IF it's a problem bundling it feel free to remove the lib inclusion requirement and I'll solve it this in the external plugin side

**Comment by thomasleveil** [comment]
It is not a problem in itself, but I now think it could be best to rework our pip-requires.txt file to maintain a list of python modules we need in different cases.

- [ ] rename `pip-requires.txt` to the standard name `requirements.txt`
- [ ] update `.travis.yml` to reflect that file name change (and get rid of `--use-mirrors` which is deprecated)
- [ ] list python modules required to run the core B3 in `requirements.txt`
- [ ] create a `build-requirements.txt` file and add to it the extra modules that are only needed for 3rd party popular plugins
- [ ] add `pip install -r build-requirements.txt` to `.travis.yml`

Some reading:
- https://pip.readthedocs.org/en/1.1/requirements.html

**Comment by thomasleveil** [comment]
see danielepantaleone/big-brother-bot#5

**Comment by 82ndab-Bravo17** [comment]
Was this issue ever solved on the OSX version of Python?

http://forum.bigbrotherbot.net/installation-support/b3-still-not-responding/

**Comment by danielepantaleone** [comment]
I sent a PR against `release-1.10` branch https://github.com/BigBrotherBot/big-brother-bot/pull/298

**Comment by danielepantaleone** [comment]
It seems that I would have to remove the requirement of inclusion of the IRC library. Travis CI is having some troubles in installing it: https://travis-ci.org/danielepantaleone/big-brother-bot/jobs/50858301

**Comment by thomasleveil** [comment]
The errors are `socket.error: [Errno 104] Connection reset by peer`, maybe it is a temporary network issue between Travis servers and Pypi

**Comment by danielepantaleone** [comment]
It happens since a week and always on the IRC library, which is the last one (probably Travis CI download too many packages in too short time and the connection gets refused.....spammer Travis spammer :smile:)

**Comment by thomasleveil** [comment]
I've sent you a PR, maybe that will help with Travis issues

**Comment by danielepantaleone** [comment]
Yes, the build is passing now. I don't know if that was the problem but indeed if such modules are alreary in the python core then we won't need them there.

**Comment by thomasleveil** [comment]
http://ci.cucurb.net/view/Dist/job/B3-1.10-danielepantaleone-setuptools/21/ :white_check_mark:

**Comment by danielepantaleone** [comment]
:+1:

**Comment by thomasleveil** [comment]
I think we got everything good on this branch to go from py2exe -> cx_freeze.

Would you mind squashing all commits from this branch so that the history on the release-1.10 stays clean?

With PyCharm, go on your local `setuptools` branch
Open the `rebase...` dialog
![image](https://cloud.githubusercontent.com/assets/42300/6205709/9725b178-b578-11e4-8f65-138b85a6fef7.png)
in the `onto` field, put the ref to the remote bigBrotherBot/big-brother-bot#release-1.10 branch (can be different alias for the remote on your system)
![image](https://cloud.githubusercontent.com/assets/42300/6205724/14383bb8-b579-11e4-96b2-b3cf639ee0f4.png)
Squash the commits toguether
![image](https://cloud.githubusercontent.com/assets/42300/6205732/49d2644c-b579-11e4-92bc-6c66f65c2e31.png)
finally push on a new branch on your github account and create a pull request

----

or from the command line (assuming `b3` is the name of your remote for git@github.com:BigBrotherBot/big-brother-bot.git)

    git rebase -i b3/release-1.10

and edit the text provided with: 
![image](https://cloud.githubusercontent.com/assets/42300/6205769/17bb8028-b57a-11e4-8d61-b20fa9017866.png)

**Comment by danielepantaleone** [comment]
Sure. I wanted to make sure the compile works also on OSX just in case we want to release an app for that platform. But that can be added later.
